Take scrollbars into account when computing Dataframe dimensions

vdonato opened the pull request:
It turns out the bug in #2543 appears only when there are enough elements in a Dataframe for a scrollbar to be necessary to display all the data. In the case where we need a vertical scrollbar, the width that we provide to the `MultiGrid` we're using to render our Dataframe ends up being slightly too small due to the presence of the scrollbar. This causes a horizontal scrollbar to appear. The other case is analogous.

To fix this, we borrowed a helper function from [react-bootstrap/dom-helpers](https://github.com/react-bootstrap/dom-helpers) for getting the width of scrollbars, then used the helper to fix our rendering dimensions calculations to take scrollbars into account.

Closes #2543

vdonato committed:
Pull in scrollbarSize function from dom-helpers package

We'll be using this to fix #2543.

vdonato committed:
Take scrollbars into account when computing Dataframe dimensions

vdonato commented:
@asaini: My PR comment gives a pretty good description of the changes made, but the TL; DR version of the changes to look for vs the current behavior on the `HEAD` of `develop` are the cases where

* a dataframe has tons of columns but few rows so only a horizontal scrollbar is needed
* the analogous case with a bunch of rows and few columns

akrolsmir commented on frontend/src/components/elements/DataFrame/DataFrameUtil.tsx:
```diff
@@ -288,6 +304,7 @@ export function getWidths(
   }
 
   const elementWidth = Math.min(distributedTableTotal, containerWidth)
+  const needsHorizontalScrollbar = distributedTableTotal > containerWidth
```
Not related at all to your change, but wow getWidths() is a crazy complicated function, it makes my head hurt just looking at it. Wonder if there's some opportunity for cleanup down the line.

vdonato replied:
Yeah, certainly something we may want to look into cleaning up in the future.

akrolsmir commented on frontend/src/components/elements/DataFrame/DataFrameUtil.tsx:
```diff
@@ -99,15 +101,20 @@ export const getDimensions = (
-    width - border,
+    // Reserve enough space to render a vertical scrollbar in case we need to
+    // do so.
+    width - border - scrollbarSize(),
```
Nit: extract this to a new variable, for clarity?

`const containerWidth = width - border - scrollbarSize()`

vdonato replied:
Done!

akrolsmir commented on frontend/src/components/elements/DataFrame/DataFrameUtil.tsx:
```diff
@@ -122,14 +129,23 @@ export const getDimensions = (
     }
   }
 
+  const totalHeight = rows * rowHeight
```
Maybe a brief comment here to describe what's going on here? eg

`// Allocate space for horizontal and vertical scrollbars, if needed`

vdonato replied:
:+1:

akrolsmir reviewed (approved):
Looks great! To get the snapshot tests to pass, you'll want to grab the failing diff.png and run it through the [diffmazing cropper](https://www.notion.so/streamlit/Working-with-E2E-Snapshots-e8de437fb46341d6a7db11d0e4feff50), then put the new snapshot in this PR.

kmcgrady reviewed (commented):
The solution is less than ideal, but I think might have to be the path.

In terms of unit testing, here are some ideas that _might_ work:
* you can use enzyme's `mount` (or from `lib/test_util`) and utilize the [getDOMNode](https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/getDOMNode.html) and verify the offset and clientWidth are the same.
* If that doesn't work, the e2e test should be able to resolve this as well, both in a snapshot test (the test should not display scrollbars) and doing the same check for `offsetWidth` and `clientWidth`

vdonato committed:
Make some small readability changes to address PR comments

vdonato committed:
Add unit tests and update snapshots

vdonato reviewed (commented):
Thanks for the reviews @akrolsmir, @kmcgrady! May be worth taking a second glance at things to make sure the tests seem sane.

kmcgrady commented:
Looks reasonable, just got to update the snapshot for the cypress test. Let me know if you need help with that.

And then @asaini should take a look at it for product review.

vdonato committed:
Update more snapshots since cypress seems to flag them one at a time

vdonato committed:
Continue updating snapshots

vdonato committed:
Continue to slowly swap out snapshots (2 left?)

At least this is a pretty strong incentive to fix the currently-broken Docker
image for running cypress :neutral_face:

vdonato committed:
Fix diff and delete those remaining to force a regen

vdonato committed:
Re-add last fixed snapshot

asaini commented:
LGTM 👍
